File: CSC454HW5/CppImpOfHW4/VendingEventModel.cpp
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

/* 
 * File:   VendingEventModel.cpp
 * Author: chrisrk192
 * 
 * Created on November 7, 2017, 7:22 PM
 */

#include "VendingEventModel.h"
#include <cmath>
#include <string>
#include <vector>
#include <iostream>
VendingEventModel::VendingEventModel() {
    numQuarters = 0; numNickels = 0; numDimes = 0; totalVal = 0;
    e = 0.0;
}

VendingEventModel::VendingEventModel(const VendingEventModel& orig) {
}

VendingEventModel::~VendingEventModel() {
}

std::vector<std::string> VendingEventModel::OutPut(){
    std::string output = "";
        //if(cancelPushed){            
        if(totalVal >= 100) {
            for(int j = 0; j < totalVal / 100; j++){
                if(j == (totalVal / 100)-1){
                    output+= EnumToString(coffee)+", ";
                } else{
                    output+= EnumToString(coffee)+", ";
                }
            }
                
        }
        std::vector<int> *l = CalcChange(totalVal%100, numQuarters, numDimes, numNickels);
        for(int i = 0; i < l->size();i++){
            if(i == l->size()-1){
                output+= EnumToString(l->at(i));
            } else{
                output+= EnumToString(l->at(i))+", ";
            }
        }
        delete l;
        std::vector<std::string> so = SplitString(output,", ");
        return so; 
}   

void VendingEventModel::StateTransitionExt(std::vector<std::string> inputItems, double ee){
    e = ee;
    ioSet s;

    s = static_cast<ioSet>(StringToEnum(inputItems[n]));

    switch(s){
        case q:
            numQuarters++;
            totalVal+=25;
            break;
        case n:
            numNickels++;
            totalVal+=5;
            break;
        case d:
            numDimes++;
            totalVal+=10;
            break; 

    }
    e = 0.0;
}

void VendingEventModel::StateTransitionInt(){
    std::vector<int> *l = CalcChange(totalVal, numQuarters, numDimes, numNickels); 
            
            for(int i = 0; i < l->size();i++){
                switch(l->at(i)){
                    case q:
                        numQuarters--;
                        totalVal-=25;
                        break;
                    case n:
                        numNickels--;
                        totalVal-=5;
                        break;
                    case d:
                        numDimes--;
                        totalVal-=10;
                        break;                        
                }
            }
            delete l;            
            if(totalVal > 0){
                CantPerformEventAction(std::string("Not enough Change."));//"Cant make all change.");
            }
}

void VendingEventModel::StateTransitionCon(std::vector<std::string> inputItems){
    StateTransitionInt();
    StateTransitionExt(inputItems,2.0);
}

double VendingEventModel::TimeAdvance(){
    if(totalVal > 0)
        return 2.0;
    return INFINITY;
}


std::string VendingEventModel::GetStatus(){
    std::string s =
        "number of quarters = " +std::to_string(numQuarters)+"\n"
                + "number of dimes    = " +std::to_string(numDimes)+"\n"
                + "number of nickels  = " +std::to_string(numNickels)+"\n"
                + "value in machine   = " +std::to_string(totalVal)+"\n"
                ;
    return s;
}

std::vector<int>* VendingEventModel::CalcChange(int value, int numQ, int numD, int numN){
    std::vector<int> *l = new std::vector<int>;
        bool changeMadeFlag = false;
        while(value > 0){
            if(value >= 25 && numQ > 0){
                l->push_back(q);
                value -= 25;
                numQ--;
                changeMadeFlag = true;
            }
            if(value >= 10 && numD > 0){
                l->push_back(d);
                value -= 10;
                numD--;
                changeMadeFlag = true;
            }
            if(value >= 5 && numN > 0){
                l->push_back(n);
                value -= 5;
                numN--;
                changeMadeFlag = true;
            }   
            if(numD <= 0 && numQ <= 0 && numN <= 0 || changeMadeFlag == false){
                value = 0;
                
            }
            changeMadeFlag = false;
                
        }
        return l;
}

int VendingEventModel::StringToEnum(std::string in){
    if(in.compare("coffee") == 0)
        return coffee;
    
    if(in.compare("d") == 0)
        return d;
    if(in.compare("n") == 0)
        return n;
    if(in.compare("q") == 0)
        return q;
    return -1;
    
}

std::string VendingEventModel::EnumToString(int enm){
    switch(enm){
        case coffee: return "coffee";
        
        case d: return "d";
        case n: return "n";
        case q: return "q";
        
    }
}

std::vector<std::string> VendingEventModel::SplitString(std::string line, std::string delim){
    size_t pos = 0;
    std::vector<std::string> inputs;
    while ((pos = line.find(delim)) != std::string::npos) {
            inputs.push_back(line.substr(0, pos));
            line.erase(0, pos + delim.length());
        }
    inputs.push_back(line);
    return inputs;        
}

File: CSC454HW4/CSC454HW4/src/csc454hw4/VendingEventModel.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package csc454hw4;

import java.util.ArrayList;

/**
 *
 * @author chrisrk192
 */
public class VendingEventModel implements EventModelInterface{

    static enum inputSet{n, d, q}
    static enum outputSet{n, d, q, coffee} 
    double e = 0.0;
    int qNum,nNum,dNum,vNum;
    public VendingEventModel(){
        qNum = 0;
        nNum = 0; 
        dNum = 0;
        vNum = 0;   
        
    }
    public VendingEventModel(int nq, int nd, int nn, int nv){
        qNum = nq;
        dNum = nd;
        nNum = nn;
        vNum = nv;
    }
    
    @Override
    public String[] OutPut() {
        String output = "";
        //if(vNum >= 100) {    
            // give out coffee
            for(int j = 0; j < vNum / 100; j++){
                if(j == (vNum / 100)-1){
                    output+= outputSet.coffee.name();
                } else{
                    output+= outputSet.coffee.name()+", ";
                }
            }
            //give out change
            ArrayList<outputSet> l = CalcChange(vNum % 100, qNum, dNum, nNum);
            for(int i = 0; i < l.size();i++){
                if(i == l.size()-1){
                    output+= l.get(i).name();
                } else{
                    output+= l.get(i).name()+", ";
                }
            }
            
                
        //}
        String[] so = output.split(", ");
        return so; 
    }

    @Override
    public void StateTransitionExt(String[] inputItems) throws CantPerformAction {
        e = 0.0;
        inputSet s;
        try{
            s = inputSet.valueOf(inputItems[0]);
        } catch(IllegalArgumentException ex){
            throw new CantPerformAction("Bad InputItems");
        }
        switch(s){
            case q:
                qNum++;
                vNum+=25;
                break;
            case n:
                nNum++;
                vNum+=5;
                break;
            case d:
                dNum++;
                vNum+=10;
                break; 
        }
    }
    
    @Override //give back change and or give coffee
    public void StateTransitionInt(){
        e = 0.0;
        //if(vNum >= 100) {
            vNum = vNum % 100;//give out coffee
            ArrayList<outputSet> l = CalcChange(vNum, qNum, dNum, nNum);
            //give out chnage
            for(int i = 0; i < l.size();i++){
               switch(l.get(i)){
                   case q:
                       qNum--;
                       break;
                   case d:
                       dNum--;
                       break;
                   case n:
                       nNum--;
                       break;  
               } 
            }
            vNum = 0;             
        //}        
    }
    @Override
    public void StateTransitionCon(String[] inputItems)throws CantPerformAction{
        StateTransitionInt();
        StateTransitionExt(inputItems);
    }
    
    @Override
    public double TimeAdvance(){
        if(vNum > 0)
            return 2.0;
        return Double.POSITIVE_INFINITY;
    }

    @Override
    public String GetStatus() {
        return    "number of quarters = " +qNum+"\n"
                + "number of dimes    = " +dNum+"\n"
                + "number of nickels  = " +nNum+"\n"
                + "value in machine   = " +vNum+"\n";
               
    }

    
    private ArrayList<outputSet> CalcChange(int value, int numQ, int numD, int numN) {
        ArrayList<outputSet> l = new ArrayList<>();
        boolean changeMadeFlag = false;
        while(value > 0){
            if(value >= 25 && numQ > 0){
                l.add(outputSet.q);
                value -= 25;
                numQ--;
                changeMadeFlag = true;
            }
            if(value >= 10 && numD > 0){
                l.add(outputSet.d);
                value -= 10;
                numD--;
                changeMadeFlag = true;
            }
            if(value >= 5 && numN > 0){
                l.add(outputSet.n);
                value -= 5;
                numN--;
                changeMadeFlag = true;
            }   
            if(numD <= 0 && numQ <= 0 && numN <= 0 || changeMadeFlag == false){
                value = 0;              
            }
            changeMadeFlag = false;               
        }
        return l;
    }
}
